subsys/net/ip/net_pkt.c

```diff
@@ -809,20 +815,25 @@ struct net_pkt *net_pkt_ref_debug(struct net_pkt *pkt, const char *caller,
 struct net_pkt *net_pkt_ref(struct net_pkt *pkt)
 #endif /* NET_LOG_LEVEL >= LOG_LEVEL_DBG */
 {
-	if (!pkt) {
```

Comment:
you want to keep that test, as for unref btw or it will enable dereferencing a NULL pointer at line 821

Reply:
I did keep that test by doing
```
ref = pkt ? atomic_get(&pkt->atomic_ref) : 0;
if (!ref) {
```
This additionally catches the error of passing a pkt with reference count of 0. Since the NET_ERR message prints pkt, the two errors can be distinguished.

Reply:
> This additionally catches the error of passing a pkt with reference count of 0.

Though if this done to catch 2 errors with one error message, maybe it's ok... But I'd suggest to add code comments for such cases.

Reply:
Efficiency should not be an issue. This loop is almost never executed more than once. Only when it is interrupted by another thread/interrupt handler, it has to loop.

I can spice up this function with `unlikely()` if you want. Both `!ref` and `!atomic_cas(&pkt->atomic_ref, ref, ref + 1)` are very unlikely.